### ISSUE TITLE
Extend routing.xml to represent the routing cost of intersections

### ISSUE THREAD

**Issue by david-gpu** [osmandapp/OsmAnd#21050]
### Describe the idea

In the real world, routing quality is highly dependent on the street intersections found along the way, and a big factor in that is the kind of `highways` that enter/exit that intersection. For example, an intersection between two residential streets is very different from an intersection between a residential road and a multi-lane secondary road, especially when the user is cycling. And whether the intersection has traffic lights or not is also very important.

As of today, [`routing.xml` has a section called `obstacle`](https://github.com/osmandapp/OsmAnd-resources/blob/bfa257b412ecc843ec38a08cdf43826d0e8180ae/routing/routing.xml#L1013) dedicated to tracking the cost associated with `points`. However, it only takes into account the tags associated with that `point`, ignoring completely any `highways` that enter/exit that `point`.

### Expected behaviour

It would be very useful to add a new section to `routing.xml` dedicated explicitly to the cost of crossing an intersection. This section would be semantically similar to the existing `obstacle` section, containing a sequence of `select` elements, but instead of matching based on the tags of the `point`, it would match based on the type of `highway` we are navigating through and the type of `highway` that intersects our route.

The `select` element could contain the following attributes:
* `value` (mandatory): The routing cost associated with this intersection, measured in seconds.
* `t` (optional; default=match): Key of tag belonging to this point.
* `v` (optional; default=match): Value of tag associated with key above.
* `nav` (optional; default=match): Type of `highway` that the route is currently navigating through. E.g. `residential` means we are currently navigating through a way tagged as `highway=residential`.
* `int` (optional; default=match): At least one of the ways that enters/exits this intersection (other than `nav`) is tagged as `highway=<int>`.

Here's an arbitrary example that would improve cycling/micromobility safety:
```
<point attribute="intersection">
    <select value="2" int="residential"/> <-- When our route intersects with a residential road, things are easy -->
    <select value="20" t="highway" v="traffic_signals"/> <!-- An intersection with traffic signals is fairly safe -->
    <select value="100" int="secondary"/> <!-- An unprotected intersection with a secondary road is undesirable -->
...
```

### Alternatives you've considered

Perhaps an alternative way of representing the cost of traversing an intersection would be changing the semantics of the [`penalty_transition`](https://github.com/osmandapp/OsmAnd-resources/blob/bfa257b412ecc843ec38a08cdf43826d0e8180ae/routing/routing.xml#L199) section, so that the penalty is applied not only when the navigation route enters a particular way, but also when the navigation route crosses an intersection with that type of way.

However, we still need to differentiate between intersections with different levels of safety. For example, an intersection _without_ traffic lights is not as safe as an intersection _with_ traffic lights.

**Comment by vshcherb** [comment]
There will be definitely some challenges with bidirectional search of A*. Probaly it should be solved differently as introducing artifical points at intersection so penalty could be applied to them. Probablly these could be enhanced with penalty_transition

**Comment by david-gpu** [comment]
You make a good point about bidirectional search -- I didn't think of that.

Let's use the specific example that I ran into. In [this map](https://www.openstreetmap.org/#map=19/43.774771/-79.390096) you can see two different ways to travel West to East:
# From Kingsdale Ave, across the unprotected intersection with Bayview Ave, towards Wycliffe Cr.
# From Empress Ave, across the traffic lights at Bayview Ave, towards Citation Dr.

![image](https://github.com/user-attachments/assets/7a709592-c61d-4dde-a2f7-1f0cd8ad2b1a)

From personal experience, and from the Strava Global Heatmap, it is very clear that the first option is almost suicide and the second option is safe, thanks to the traffic lights. Even cars avoid the first option, even though it is (surprisingly) legal.

I like the idea of creating virtual points to represent intersections. It simplifies the 1-to-N relationship to multiple 1-to-1 relationships, and I think it takes care of the bidirectional routing algorithm as well: you only need to swap the "to" and "from" ways.

Another benefit of these virtual points would be handing the fact that in many parts of the USA and Canada vehicles are allowed to turn right while their traffic light is still red (i.e. _right turn on red_), so the real-world routing cost of going through an intersection depends on which direction you turn. Each of the options could be represented with a separate virtual point, where the penalty of a right turn on red would be lower than the other options.

**Comment by david-gpu** [comment]
Let me clarify to make sure we are on the same page. What I'm proposing doesn't require neighboring edges penalties. Instead, the graph would be (virtually?) transformed to replace each OSM intersection (i.e. every point with three or more ways entering/exiting it).

Let me illustrate with the same example as before. This is how OSM looks now; the minor street Kingsdale Ave travels West to East and intersects with a major road Bayview Ave that travels North to South.

![image](https://github.com/user-attachments/assets/fcf65102-68e5-49e9-b134-77a186386b95)

And here is how that single intersection point would be (virtually) transformed to represent the different ways one can cross that intersection. I believe this eliminates the need for neighboring edges penalties. What we are doing is transforming an N-to-N intersection into multiple 1-to-1 edges, where each possible way to traverse the intersection can now have a different penalty:

![image](https://github.com/user-attachments/assets/d2064b8d-4712-4f97-ad99-b936a8a96a2b)

I have placed the virtual points all around so that they are visible on the map, but in the actual implementation they would all be in the same position as the original intersection.

**Comment by vshcherb** [comment]
that would be even more complicated but first of all each intersection without traffic light should have a special point that penalty could be assigned. Then it should be possible (in future) to apply reduction of penalties if they are consequent, it could be useful for traffic lights as well, however it still sounds very tricky to implement

**Comment by david-gpu** [comment]
This method doesn't introduce any neighboring edges penalties because the virtual ways that are traversed in order to travel South when you arrive from the West are different from the virtual ways that are traveled when you came from the North or the East. The code only needs to apply a different penalty depending on whether you go straight, turn left or turn right (plus u-turns).

It is not necessary to add any actual points or vertices, even in memory. They are just an abstraction to represent that when you travel from the West and turn South you are following a different "virtual way" to get to the same location.

Let me offer another way of representing these virtual ways which is basically equivalent to what I showed earlier, but perhaps easier to understand. Each virtual way represents one way of traversing the intersection: straight ahead, turn left and turn right.

![image](https://github.com/user-attachments/assets/c0c624eb-e3e4-4bc6-a899-77adff3ed38d)

**Comment by vshcherb** [comment]
Yes that's exactly penalty transition but it doesn't check adjacent ways. You want penalty for crossing secondary way without traffic signals, so the easiest way to do that introduce "virtual signs" like "stop" or "give_away" and extra edges that avoid that sign for turning right. In  that case it won't suggest that route. 
However it's still tricky to implement

**Comment by pinkshinyrose** [comment]
If this is implemented, the penalty should probably account for other features, unless it's specific to North America. Many people avoid signalled crossings because traffic lights tend to waste a lot of time. But it depends on what kind of road you're crossing. If you're crossing a 6 lane road with no island and a high speed limit, then you obviously prefer the traffic light. But for roads with one lane each way, a reasonable speed limit and a traffic island the calculus would be very different.

I'm not sure any penalty should apply if lanes=1 as it suggests a traffic island or similar crossing. For lanes=2 and (implicit) maxspeed<=60 (40 mph) you'll probably want to avoid it, but signalled or unsignalled would be about even (would probably prefer signals with a kid, unsignalled with only adults). For any larger/faster roads you'll probably want the unsignalled penalty to scale with lanes and maxspeed.

David-gpu seems to assume 5 lanes (not sure about speed limits in their area), so 5 * [1.44 for one way] * lanes^2 * maxspeed^2 / 3600 seems to make sense for unsignalled in their example (it would yield 125 if maxspeed≈40mph, which is similar to their proposal) and would be equal to their signalled penalty for a lanes=2 maxspeed=50 one-way road.

For residential streets the same system could be applied (except the lanes tag tends to be absent, but specifically for residential streets we can assume lanes=1 in those cases).
